### PR TITLE
dash/modified-table-row-relations

### PR DESCRIPTION
**Pull request by ddragula**
Added methods ([getLocalRowIndex](https://api.highcharts.com/dashboards/#classes/Data_DataTable.DataTable-1#getLocalRowIndex) & [getOriginalRowIndex](https://api.highcharts.com/dashboards/#classes/Data_DataTable.DataTable-1#getOriginalRowIndex)) for referencing between rows of original and modified data tables.

---
**CHANGES DESCRIPTION:**

Now, each DataTable modified with a modifier that changes the original row order, stores a list of the original row indexes in the original table.

Two important methods added for external use:

- `getLocalRowIndex` - takes the original row index as an argument and returns the local row index in the modified table for which this function is called.

- `getOriginalRowIndex` - takes the local row index (in the modified table for which it is called) as an argument and returns the index of the corresponding row in the original table.

_Further changes, e.g. basing the entire synchronization on these methods, will be introduced for the version with DataGrid2 (branch `dash/datagrid2`)._

---
**TO DO:**
- [x] Remove the `rowKeysId` feature to be able to replace it.
- [x] Add row references to the modified and original table.
- [x] Handle these references when used in appropriate data modifiers.
- [x] Organize tests.
- [x] Organize the code & examples.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207895050680669

### CODE REVIEW

File: ts/Data/DataTable.ts [hunk 1]
```diff
@@ -792,6 +797,28 @@ class DataTable implements DataEvent.Emitter {
         return columns;
     }
 
+    /**
+     * Takes the local row index and returns the index of the corresponding row
```
Comment (karolkolodziej) [inline ts/Data/DataTable.ts:801]: The description doesn't match the method.

File: ts/Data/DataTable.ts [hunk 2]
```diff
@@ -803,6 +830,28 @@ class DataTable implements DataEvent.Emitter {
         return this.modifier;
     }
 
+    /**
+     * Takes the original row index and returns the local row index in the
```
Comment (karolkolodziej) [inline ts/Data/DataTable.ts:834]: The description doesn't match the method.

**Review by karolkolodziej — CHANGES_REQUESTED** [review]
Do you think it makes sense to add a benchmark with some sorting to the `Benchmark report`?

**Review by jomar-honsi — APPROVED** [review]

**Review by karolkolodziej — APPROVED** [review]

File: docs/dashboards/data-modifiers.md [hunk 3]
```diff
@@ -93,6 +93,13 @@ dataModifier: {
 ```
 [API documentation](https://api.highcharts.com/dashboards/typedoc/interfaces/Data_Modifiers_SortModifierOptions.SortModifierOptions-1.html)
 
+### Original and Modified Table relations
+
+Each modified table contains two methods that allow you to manage the relationships between the rows of the original and modified tables:
```
Comment (sebastianbochan) [inline docs/dashboards/data-modifiers.md:98]: Every (instead of each)

**Review by sebastianbochan — APPROVED** [review]

File: ts/Data/DataTable.ts [hunk 4]
```diff
@@ -346,15 +346,9 @@ class DataTable implements DataEvent.Emitter {
...
-                // All columns deleted, remove row keys column
-                delete columns[table.rowKeysId];
-                nColumns = 0;
-            }
-
-            if (!nColumns) {
+            if (!Object.keys(columns).length) {
                 table.rowCount = 0;
+                this.deleteRowIndexReferences();
             }
```
Comment (bre1470) [inline ts/Data/DataTable.ts:352]: Maybe it is time for a `DataTable.clear` function, that keeps columns and only removes rows and index refs.

**Review by bre1470 — APPROVED** [review]
Awesome work! Looks solid. 👍

File: ts/Data/DataTable.ts (hunk 4 above)
Comment (ddragula) [reply ts/Data/DataTable.ts:352]: How would it differ from `DataTable.deleteRows`?

I realized that I forgot about this method and didn't implement the references updating if we use `deleteRows` on a modified table.  I'll take care of it right away.

**Pull request merged**